crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2178 @@
+ * then this code will also need change.
+ */
+static int hpke_random_suite(OSSL_LIB_CTX *libctx,
+                             const char *propq,
+                             OSSL_HPKE_SUITE *suite)
+{
+    unsigned char rval = 0;
+    int nkdfs = OSSL_NELEM(hpke_kdf_tab)-1;
+    int naeads = OSSL_NELEM(hpke_aead_tab)-1;
+    int nkems = OSSL_NELEM(hpke_kem_tab)-1;
+
+    /* random kem */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->kem_id = hpke_kem_tab[(rval % nkems + 1)].kem_id;
+
+    /* random kdf */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->kdf_id = hpke_kdf_tab[(rval % nkdfs + 1)].kdf_id;
+
+    /* random aead, but not exportonly */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->aead_id = hpke_aead_tab[(rval % (naeads - 1) + 1)].aead_id;
+    return 1;
+}
+
+/*
+ * @brief return a (possibly) random suite, public key, ciphertext for GREASErs
+ *
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param suite-in specifies the preferred suite or NULL for a random choice
+ * @param suite is the chosen or random suite
+ * @param pub a random value of the appropriate length for sender public value
+ * @param pub_len is the length of pub (buffer size on input)
+ * @param ciphertext buffer with random value of the appropriate length
+ * @param ciphertext_len is the length of cipher
+ * @return 1 for success, otherwise failure
+ */
+static int hpke_get_grease_value(OSSL_LIB_CTX *libctx, const char *propq,
+                                 OSSL_HPKE_SUITE *suite_in,
+                                 OSSL_HPKE_SUITE *suite,
+                                 unsigned char *pub,
+                                 size_t *pub_len,
+                                 unsigned char *ciphertext,
+                                 size_t ciphertext_len)
+{
+    OSSL_HPKE_SUITE chosen;
+    int crv = 0;
+    int erv = 0;
+    size_t plen = 0;
+    uint16_t kem_ind = 0;
+
+    if (pub == NULL || !pub_len
+        || ciphertext == NULL || !ciphertext_len || suite == NULL)
+        return 0;
+    if (suite_in == NULL) {
+        /* choose a random suite */
+        crv = hpke_random_suite(libctx, propq, &chosen);
+        if (crv != 1)
+            return crv;
+    } else {
+        chosen = *suite_in;
+    }
+    kem_ind = kem_iana2index(chosen.kem_id);
+    if (kem_ind == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if ((crv = hpke_suite_check(chosen)) != 1)
+        return 0;
+    *suite = chosen;
+    /* publen */
+    plen = hpke_kem_tab[kem_ind].Npk;
+    if (plen > *pub_len)
+        return 0;
+    if (RAND_bytes_ex(libctx, pub, plen, OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    *pub_len = plen;
+    if (RAND_bytes_ex(libctx, ciphertext, ciphertext_len,
```

Comment:
Not sure if this should use the _priv variant of this API for the ciphertext.

Reply:
For GREASE, it's better to not use the _priv variant I think - if there's a real difference between the random streams from the two variants, then I'd say better to use the more "public" stream (as used for e.g. message IDs or whatever) for GREASE values as those are directly visible to the network. (So if e.g. some implementation were using a borked PRNG or dual-ec type thing the GREASE values might be less likely to expose the state of the _priv variant PRNG.)